readthedocs/vcs_support/backends/git.py

```diff
@@ -23,6 +24,12 @@
 log = structlog.get_logger(__name__)
 
 
+@dataclass(slots=True)
+class GitSubmodule:
+    path: str
+    url: str
```

Comment:
We need the path and URL of each submodule, since we validate that each URL has a valid/supported schema, we could remove this check if we want, it would simplify the code a little :)

Reply:
And we would just need to call `Git.submodules` when the user is excluding submodules only, when including some or all, we can just call the `git submodule update` with an empty list or the list of given submodules.

Reply:
And if the user provides submodules that aren't valid, well, git will just take care of showing the correct error message, right now we are supressing those.

Reply:
@agjohnson would you be +1 on removing that check and let git just error if the user is trying to use ssh/invalid URLs?

Reply:
https://github.com/readthedocs/readthedocs.org/blob/8cae85b14d2dbc99300f17d33a493922ef0481c1/readthedocs/vcs_support/backends/git.py#L337-L341

Reply:
Thinking about this more, I seem to recall some issues with repositories defining bogus submodules, or having old and unused submodules defined. I don't think this was to limit the URL pattern to supported protocols.

It could have also been a security measure, as it was probably possible to define a submodule like `/etc/` or something and clone from something outside the container (when we were performing submodule operations outside containers).

Any of this seem familiar or likely?

Reply:
Stealing internal repos could have been a possibility before moving the execution to docker.

Reply:
It sounds familiar to me the validation of the submodules, but this PR should fix that automatically because we trust git itself now

Reply:
Yeah, I'm thinking if any of the above are true, it was probably a throw back to when we weren't executing the submodule steps inside Docker. In any case that would seem this logic can be removed, but if anyone is hesitant about it, I'm not saying this is guaranteed to be safe either.

Reply:
In that case, I'd say this is up to you @stsewd since you are the security master here 😄